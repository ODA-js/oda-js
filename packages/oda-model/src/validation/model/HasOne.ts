import { IRelationPropsStore } from '../interfaces/IRelation';
import { Record } from 'immutable';
import { Map, Set } from 'immutable';

import { Persistent } from './Persistent';
import { transformMap, transformSet } from './utils';
import { IHasOneRelationPropsStore, IHasOneRelationProps, IHasOneRelation } from '../interfaces/IHasOneRelation';
import { IEntityRef } from '../interfaces/IEntityRef';
import { IField } from '../interfaces/IField';

// tslint:disable-next-line:variable-name
export const DefaultHasOne: IHasOneRelationPropsStore = {
  modelType: 'relation',
  name: null,
  title: null,
  description: null,
  verb: null,
  ref: null,
  fields: null,
  opposite: null,
  hasOne: null,
};

// tslint:disable-next-line:variable-name
export const RelationTransform: { [ k in keyof IHasOneRelationPropsStore]?: any }  = {
  fields: transformMap<IField>(),
};

// tslint:disable-next-line:variable-name
export const HasOneStorage = Record(DefaultHasOne);

export class HasOne extends Persistent<IHasOneRelationProps, IHasOneRelationPropsStore> implements IHasOneRelation {
  public get modelType(): 'relation' {
    return 'relation';
  }
  public get name(): string {
    return this.store.get('name', null);
  }
  public get title(): string {
    return this.store.get('title', null);
  }
  public get description(): string {
    return this.store.get('description', null);
  }
  public get ref(): IEntityRef {
    return this.store.get('ref', null);
  }
  public get fields(): Map<string, IField> {
    return this.store.get('fields', null);
  }
  public get opposite(): string {
    return this.store.get('opposite', null);
  }

  public get verb(): 'HasOne' {
    return 'HasOne';
  }
  public get hasOne(): IEntityRef {
    return this.store.get('hasOne', null);
  }

  protected transform(input: IHasOneRelationProps): IHasOneRelationPropsStore {
    return {
      ...input,
      fields: RelationTransform.fields.transform(input.fields),
    };
  }
  protected reverse(input: IHasOneRelationPropsStore): IHasOneRelationProps {
    return {
      ...input,
      fields: RelationTransform.fields.reverse(input.fields),
    };
  }
  constructor(init: IHasOneRelationProps) {
    super();
    this.store = new HasOneStorage(this.transform(init));
    this.init = new (Record<IHasOneRelationProps>(init))();
  }
}
