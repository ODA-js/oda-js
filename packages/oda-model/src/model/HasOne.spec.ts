import { HasOne } from './HasOne';

describe('HasOne', () => {
  let relation: HasOne;

  beforeAll(() => {
    expect(() => new HasOne()).not.toThrow();
  });

  beforeEach(() => {
    relation = new HasOne();
  });

  it('create empty', () => {
    expect(relation.modelType).toBe('relation');
    expect(relation.verb).toBe('HasOne');
    expect(relation.title).toBeNull();
    expect(relation.name).toBeNull();
    expect(relation.description).toBeNull();
    expect(relation.fields).toBeNull();
    expect(relation.fullName).toBeNull();
    expect(relation.shortName).toBeNull();
    expect(relation.normalName).toBeNull();
    expect(relation.opposite).toBeNull();
    expect(relation.embedded).toBeFalsy();
    expect(relation.single).toBeTruthy();
    expect(relation.stored).toBeFalsy();
    expect(relation.ref).toBeNull();
    expect(relation.hasOne).toBeNull();
  });

  it('updates strings', () => {
    expect(() => relation.updateWith({
      name: 'cool',
      description: 'very cool',
      title: 'very cool title',
    })).not.toThrow();
    expect(relation.name).toBe('cool');
    expect(relation.description).toBe('very cool');
    expect(relation.title).toBe('very cool title');
    expect(() => relation.updateWith({
      name: 'cool!',
    })).not.toThrow();
    expect(relation.name).toBe('cool!');
  });
  it('toJS with dupes', () => {
    expect(() => relation.updateWith({
      fields: [
        { name: 'one' },
        { name: 'one' },
      ],
    },
    )).not.toThrow();

    expect(relation.ref).toBeNull();
    expect(relation.fields.size).toBe(1);
    expect(relation.toJS()).toMatchObject({
      fields: [
        { name: 'one' },
      ],
    });
  });

  it('toJS ', () => {
    expect(() => relation.updateWith({
      hasOne: 'id@Some#id',
    },
    )).not.toThrow();

    expect(relation.ref).not.toBeNull();
    expect(relation.ref.backField).toBe('id');
    expect(relation.ref.entity).toBe('Some');
    expect(relation.ref.field).toBe('id');

    expect(relation.ref.toJS()).toMatchObject({
      entity: 'Some',
      backField: 'id',
      field: 'id',
    });

    expect(relation.toJS()).toMatchObject({
      hasOne: 'id@Some#id',
    });
  });
});
