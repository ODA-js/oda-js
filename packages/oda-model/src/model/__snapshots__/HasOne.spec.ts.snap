// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HasOne create empty 1`] = `
HasOne {
  "store": Immutable.Record {
    "name": null,
    "title": null,
    "description": null,
    "fields": null,
    "opposite": null,
    "hasOne": null,
    "single": true,
    "stored": false,
    "embedded": false,
    "fullName": null,
    "normalName": null,
    "shortName": null,
  },
}
`;

exports[`HasOne create empty 2`] = `
Object {
  "embedded": false,
  "single": true,
  "stored": false,
}
`;

exports[`HasOne toJS  1`] = `
Object {
  "backField": "id",
  "entity": "Some",
  "field": "id",
}
`;

exports[`HasOne toJS  2`] = `
Object {
  "embedded": false,
  "hasOne": "id@Some#id",
  "single": true,
  "stored": false,
}
`;

exports[`HasOne toJS with HashMap 1`] = `
Object {
  "embedded": false,
  "fields": Array [
    Object {
      "name": "one",
    },
  ],
  "single": true,
  "stored": false,
}
`;

exports[`HasOne toJS with dupes 1`] = `
Object {
  "embedded": false,
  "fields": Array [
    Object {
      "name": "one",
    },
  ],
  "single": true,
  "stored": false,
}
`;

exports[`HasOne updates strings 1`] = `
Object {
  "description": "very cool",
  "embedded": false,
  "name": "cool",
  "single": true,
  "stored": false,
  "title": "very cool title",
}
`;

exports[`HasOne updates strings 2`] = `
Object {
  "description": "very cool",
  "embedded": false,
  "name": "cool!",
  "single": true,
  "stored": false,
  "title": "very cool title",
}
`;

exports[`HasOne updates with null or undefined 1`] = `
Object {
  "description": "very cool",
  "embedded": false,
  "name": "cool",
  "single": true,
  "stored": false,
  "title": "very cool title",
}
`;

exports[`HasOne updates with null or undefined 2`] = `
Object {
  "embedded": false,
  "name": "cool",
  "single": true,
  "stored": false,
}
`;
